pkg/resto/registry.go

```diff
@@ -0,0 +1,157 @@
+package resto
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/registry/client/auth"
+	"github.com/docker/distribution/registry/client/transport"
+	"github.com/docker/docker/api/types"
+	dd "github.com/docker/docker/distribution"
+	"github.com/docker/docker/registry"
+	digest "github.com/opencontainers/go-digest"
+)
+
+type myreference string
+
+func (m myreference) String() string {
+	return string(m)
+}
+
+func (m myreference) Name() string {
+	return string(m)
+}
+
+// MediaTypeConfig is the media type used for configuration files.
+const MediaTypeConfig = "application/vndr.docker.config"
+
+// ConfigManifest is a Manifest type holding arbitrary data.
+type ConfigManifest struct {
+	mediaType string
+	payload   []byte
+}
+
+// References returns the objects this Manifest refers to.
+func (c *ConfigManifest) References() []distribution.Descriptor {
+	return nil
+}
+
+// Payload returns the mediatype and payload of this manifest.
+func (c *ConfigManifest) Payload() (string, []byte, error) {
+	return c.mediaType, c.payload, nil
+}
+
+// NewConfigManifest creates and returns an new ConfigManifest.
+func NewConfigManifest(mediaType string, payload []byte) *ConfigManifest {
+	return &ConfigManifest{mediaType, payload}
+}
+
+func init() {
+	distribution.RegisterManifestSchema(MediaTypeConfig, func(b []byte) (distribution.Manifest, distribution.Descriptor, error) {
+		return &ConfigManifest{
+				mediaType: MediaTypeConfig,
+				payload:   b,
+			},
+			distribution.Descriptor{
+				MediaType: MediaTypeConfig,
+				Size:      int64(len(b)),
+				Digest:    digest.SHA256.FromBytes(b),
+			}, nil
+	})
+}
+
+// NewRepository instantiates a distribution.Repository pointing to the given target, with credentials
+func NewRepository(ctx context.Context, endpoint string, repository string, opts RegistryOptions) (distribution.Repository, error) {
+	named := myreference(repository)
+	authConfig := &types.AuthConfig{
+		Username: opts.Username,
+		Password: opts.Password,
+	}
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	apiendpoint := registry.APIEndpoint{
+		Mirror:    false,
+		URL:       url,
+		Version:   2,
+		TLSConfig: &tls.Config{InsecureSkipVerify: opts.Insecure},
+	}
+	repoInfo, err := registry.ParseRepositoryInfo(named)
+	if err != nil {
+		return nil, err
+	}
+	repo, _, err := dd.NewV2Repository(ctx, repoInfo, apiendpoint, nil, authConfig, "push", "pull")
+	if err == nil {
+		return repo, nil
+	}
+	if !strings.Contains(err.Error(), "HTTP response to HTTPS client") {
+		return repo, err
+	}
+	if os.Getenv("RESTO_INSECURE_AUTH") != "1" {
```

Comment:
This looks to be  global configuration, which means that Docker Hub can be MITM'd; For comparison, the docker daemon requires each registry that is allowed to be accessed insecurely to be configured (through the  `--insecure-registry` option).

For the `cli`, connecting to an insecure registry, there should be an option to set it for each action (e.g., similar to how `docker manifest inspect` has an `--insecure` flag)

Reply:
We are not a daemon though. By default docker-app/resto won't transmit credentials over an insecure channel unless this override environment variable is used.

Reply:
So, what will happen is that someone sets `RESTO_INSECURE_AUTH=1` to test against a local registry, but now opens a major security hole by allowing insecure access _all_ registries

Reply:
That seems an unlikely scenario.

Reply:
So is people doing `chmod 777 /var/run/docker.sock`, or `sudo rm -rf /var/lib/docker`

Reply:
I don't particularly intend to document this env variable, I see it more as an escape hatch in the unlikely scenario one user end up needing it. If that ever happens we will make sure to emphasis the advice to pass it directly on the command line and not to export it or god forbid add it to a .rc file.

Reply:
It's not an unlikely scenario https://github.com/moby/moby/issues/8887, and it _will_ happen; add this escape hatch, and there'll be no turning back